src/ui/index.test.ts

```diff
@@ -1,25 +1,317 @@
+        {
+          group: "default",
+          attributes: {
+            name: "a",
+            type: "checkbox",
+            node_type: "input",
+          },
+        },
+        {
+          group: "foo",
+          attributes: {
+            name: "c",
+            type: "hidden",
+            node_type: "input",
+          },
+        },
+        {
+          group: "default",
+          attributes: {
+            name: "g",
+            type: "hidden",
+            node_type: "input",
+          },
+        },
+      ],
+    },
+    {
+      description:
+        "can filter by nodes using a group and exclude default groups",
+      opts: {
+        groups: ["foo"],
+        withoutDefaultGroup: true,
+      },
+      expected: [
+        { group: "foo", type: "b" },
+        {
+          group: "foo",
+          attributes: {
+            name: "c",
+            type: "hidden",
+            node_type: "input",
+          },
+        },
+      ],
+    },
+    {
+      description: "can filter by attributes and exclude default attributes",
+      opts: {
+        attributes: "hidden",
+        withoutDefaultAttributes: true,
+      },
+      expected: [
+        {
+          group: "foo",
+          attributes: {
+            name: "c",
+            type: "hidden",
+            node_type: "input",
+          },
+        },
+        {
+          group: "default",
+          attributes: {
+            name: "g",
+            type: "hidden",
+            node_type: "input",
+          },
+        },
+      ],
+    },
+    {
+      description: "can filter by attributes and exclude another attributes",
+      opts: {
+        groups: "webauthn",
+        withoutDefaultGroup: true,
+        excludeAttributes: "script",
+      },
+      expected: [
+        {
+          group: "webauthn",
+          attributes: {
+            name: "e",
+            type: "input",
+            node_type: "input",
+          },
+        },
+      ],
+    },
+    {
+      description:
+        "filtering by attributes should always include the default attributes even when group is not defined",
+      opts: {
+        attributes: "something",
+      },
+      expected: [
+        {
+          group: "webauthn",
+          attributes: {
+            name: "d",
+            type: "script",
+            node_type: "input",
+          },
+        },
+        {
+          group: "webauthn",
+          attributes: {
+            name: "e",
+            type: "input",
+            node_type: "input",
+          },
+        },
+        {
+          group: "totp",
+          attributes: {
+            name: "f",
+            type: "input",
+            node_type: "input",
+          },
+        },
+        {
+          group: "default",
+          attributes: {
+            name: "g",
+            type: "hidden",
+            node_type: "input",
+          },
+        },
+      ],
+    },
+  ]
+
+  tc.forEach(({ description, opts, expected }) => {
```

Comment:
Jest has direct support for table tests. No need to rewrite it here, but just so you know ;)

https://jestjs.io/docs/api#testeachtablename-fn-timeout

Reply:
this is quite insightful :)